### ISSUE TITLE
allow "depends on" dataset

### ISSUE THREAD

**Issue by DrewDennison** [returntocorp/inputset-generator#18]
@inkz wants to create an input set for all npm/python packages that depend on a certain dependance. e.g. all packages that depend on `lodash` or something

**Comment by inkz** [comment]
yep, as long as there are different `input_type`s it would be also cool to be able to specify whether I want to generate a list of npm packages or its GitHub repositories
(if it is possible)

**Comment by mschwager** [comment]
Looks like the Github API doesn't support querying for repository dependent information yet. But we may be able to use something like https://github.com/nickcolley/github-dependents-api, which scrapes the web pages.

**Comment by inkz** [comment]
@mschwager Oh yeah I understand that Github currently does not provide the opportunity for querying dependents, but speaking of my personal expectations of this feature, I meant something like this:

1. fetching info about npm dependants
2. filtering those that have Github repo in the description
3. then filtering those repos that actually still exist (this may be omitted, but I used this step in order to reduce error rate on the platform)

this is the approach that I used while crafting my own scripts for inputset generation, of course it does not give the full list of all possible dependants, but still provides enough research material, which is good enough for me.

**Comment by mschwager** [comment]
After further investigation here's what I found. All these services typically break projects down into 3 categories: PyPI package, NPM package, and Github repository. Dependents, dependencies, etc will often refer to these categories of projects.

There are many ways to attempt to gather this information:

- https://github.com/nickcolley/github-dependents-api
  - This seems like useful API built on the Github's dependent information, but in my testing it tends to choke on larger response sizes (e.g. packages with may dependents).
- Scrape the information from Github ourselves
  - This would probably fix the above issue, but it seems like a fragile solution.
- Scrape the information from NPM and PyPI
  - Looks like this possible for NPM via https://www.npmjs.com/browse/depended/<package>
    - `curl -H 'X-Spiferack: 1' https://www.npmjs.com/browse/depended/react`
  - Though we'd have to scrape all dependents then grab their repository by scraping their package information
  - I don't think it's possible to scrape Python dependents via PyPI - looks like they don't track this info
  - May be possible through BigQuery? https://packaging.python.org/guides/analyzing-pypi-package-downloads/ 
- https://api-docs.npms.io/
  - Includes package info, and dependent count, but not dependent info
- https://libraries.io/api#project-dependents
  - This seems like it may be the best option
  - The downside is it's API key rate-limited to 60 req / min, so usage would require users to go create an API key
  - Dependents info doesn't seem to match with Github info
  - I'm not sure how the API will handle projects with lots of dependents, e.g. 10k+

The libraries.io option seems like the most reliable, although it depends on an API key. We also have to consider the 3 categories above. If we only care about NPM package information we may be able to get by with scraping NPM.

If you have your own scripts that work the best option may be to wait and, 1.) see if demand for this functionality grows, 2.) see if better options for obtaining this information come about.

**Comment by inkz** [comment]
@mschwager you totally right, scraping is not the best solution for this purpose at all, frankly speaking while discussing this feature with @DrewDennison I did not think about how it actually should work behind the curtain, just was making wishes :) sorry for taking your time on it.

I also totally forgot that I use different kinds of pre-loaded data (usually tons of package.json information) to parse it and filter out things that I need for the particular inputset,
I do not remember exactly but I guess that the last time I was running this software https://github.com/nice-registry/dependent-packages in order to get some freshly built collection of the dependents and devDependents.

**Comment by mschwager** [comment]
No worries, this is a great feature request! I would definitely like to support this kind of functionality. One huge use-case I see is searching a project's dependents for buggy misuse of that project's API. Narrowing down the input set to the project's dependents would be very useful. I'd be interested in seeing any code you were using if you have it.

Like I said, we can always wait and see if demand grows, others have any ideas for gathering this information, or services pop up that have exactly what we need (e.g. a Github API endpoint for gathering dependent information). Thanks for the feature request! I will leave it open so we can continue to track it :+1:

**Comment by bafulton** [comment]
@mschwager, a couple of thoughts:

1. Yeah, getting a list of projects dependents is not a simple task! If we do decide to go the libraries.io route though, I don't think requiring users to acquire an API key would be too unreasonable. After all, they already have to provide a github API key to not be limited to ~100 requests/hour.

2. As a halfway solution, we could add a command that converts an npm or pypi dataset into a github dataset by reading through each project's metadata and searching for github urls. The usage flow could be something like `load npm list top4kyear` --> `get --metadata` --> `gitify` --> [`get --metadata/--versions`, etc.]. We'd have to drop projects where we couldn't find a github url (and add some nice warning text), of course.

    With this approach, @inkz would only have to hand-generate the initial list of dependent projects, and then he could dump it into ISG as an npm/pypi file and go from there.

    We might even be able to take advantage of the `meta` variable in each project to automatically define a pointer to the github url for each npm/pypi project when we're generating them, which would then make conversion quite simple. (This paragraph's a little hand-wavey; I'd need to think through it and provide a better explanation if we go this route.)

3. Right now, API objects are linked to specific datasets, and it's one API per dataset. This was a decision of convenience at the time, but it seems a little incongruous now. If we do go the route of creating a libraries.io API, suddenly we could have multiple APIs associated with a single dataset. We should probably consider linking API objects to the session, rather than the dataset. This would require some refactoring, but may make sense. Thoughts for the future. 🙂

**Comment by inkz** [comment]
speaking of libraries.io API, I tried it recently and it seems like `dependents` endpoint doesn’t work,

if you request the one from the example it is fine
https://libraries.io/api/NPM/base62/dependents?api_key=

but if you try something else like
https://libraries.io/api/NPM/express/dependents?api_key=

it hangs up :confused:
